### PR TITLE
Make the package hash consistent

### PR DESCRIPTION
**Pull request by tgamblin**
This PR ensures that the package hash is consistent across Python versions, and adds a lot of tests to ensure that.  It also makes the package hash less sensitive, to avoid superfluous rebuilds in CI.

Now that we have #25310 (reusing installed packages), we want to make Spack hashes more fine-grained to preserve provenance.  To do that we want to switch to using full hashes everywhere.  Full hashes include all build dependencies, a hash of the canonicalized package file, and the hashes of artifacts and patches used in the build.

The existing full hash worked by editing the AST of a `package.py` file, removing multimethods (where statically possible b/c we know they won't be triggered), and stripping comments, docstrings, and directives.  Comments and docstrings don't affect the build, and directives are represented by other metadata on the spec.  Once all these modifications were done, we dumped the `repr()` of the AST and hashed that.

There were a few problems with the old package hash:
1. The AST changes pretty frequently from Python version to Python version, and so does its `repr()`.  So, Spack running under Python 2.7, 3.5, and 3.8 may calculate entirely different full hashes for the same `package.py` file.  This isn't as big a deal *right now*, but it needs to be fixed if we are going to move to full hashes.
2. Static multimethod reosolution (removing multi-methods if we know they won't be used) was not working in all cases in the previous version of the hash, and there weren't a lot of tests for it.  This resulted in the old package hash stripping *any* function with a decorator from the canonical representation, so we would lose some functions that affected builds.
3. The old docstring remover would only strip the first free-standing string in a module/class/method.  We want to strip all of them (a few packages had extra strings lying around).
4. The old package hash would not remove directive nested in complex logic at the package level (`if`, `for`, `with when()`, etc.), and that constitutes a lot of logic for some packages, even when all the loops are doing is setting up complex directives.  
5. The old package hash was considering changes to `maintainers` and `tags`, and we really don't want those to require a rebuild.  

This PR makes the following modifications:

- [x] To solve the AST issue, add `spack.util.unparse`, which regenerates Python source code from a parsed AST.  This is based on https://github.com/simonpercivall/astunparse, which is consistent across Python 2.7-3.8, with back ports from the new `ast.unparse` function in Python 3.9+.
    * The `unparse()` method implemented here is consistent for Python 2.7 and 3.5-3.10.
    * We added a `py_ver_consistent=True` argument to `unparse()`, which handles the following cases:
         * Ensure that `*args` and `**kwargs` are last in function calls, for consistency with the Python 2 AST (which has no idea what position they're in)
         * Always unparse `print()` as a function (Python 2's AST treats it as its own statement) and do *not* differentiate between `print(a, b, c)` and `print((a, b, c))`.  These have different semantics in Python 2 and Python 3, but they almost certainly do not affect the results of builds, so switching between them won't affect the package hash.
         * Unparse unicode literals the way Python 2 would.  Basically, this means that all non-ASCII characters will use `\x` encoding in output, and unicode literals will be unparsed with a `u` prefix.
    * Our `unparse()` is consistent with `ast.unparse()` when `py_ver_consistent` is not used, so that we *may* be able to switch to the builtin `ast.unparse()` one day (if we ever only support Python 3.9+).  This is done with backports of things like `ast.unparse`'s parenthesis precedence algorithm.
    * Add some tests that guarantee that packages will have the same hash across different python versions.
- [x] Fix multimethod resolution in `package_hash.py` to preserve non-`when` decorators, and handle dynamic `@when` statements better. Add a bunch of tests for this.
- [x] Fix docstring stripping to remove any free-standing string in a `package.py` file.  This is safe as these strings are used as comments -- they don't affect build results.
- [x] Add some new commands to deal with package hashes:
    * `spack pkg source <spec>`: dump source for a `package.py` file
    * `spack pkg source --canonical <spec>`: dump canonical source for a `package.py` file, with comments and docstrings stripped and multimethods removed.
    * `spack pkg hash <spec>`: print the canonical package hash for the spec.
- [x] Remove the old AST-dump-based `package_hash` and use the new, stable `unparse`-based one.
- [x] New unparser descends into control logic a the package level and removes anything (loops, ifs, with statements) that is comprised only of directives.  This makes the package hash less sensitive.
- [x] Added more attributes for removal (`maintainers`, `tags`) to make the hash less sensitive.

I've verified that this is consistent for Python 2.7 and 3.5-3.10 across all Spack packages by unparsing them and comparing their canonicalized source.  The test packages in `lib/spack/spack/test/data/unparse` and the `test_package_hash_consistency()` test are meant to preserve this with some spot checks of interesting cases.

This should trigger a large CI rebuild on *this PR*, but it may reduce CI churn in the future (though I think CI was only using one version of Python, so maybe not).

### CODE REVIEW

File: lib/spack/external/spack_astunparse/unparser.py
Comment (scheibelp) [inline lib/spack/external/spack_astunparse/unparser.py]: (minor) this is a typo (no space between "character" and "':'". More importantly, I'm curious if this comment will be displayed in the PR (I made this comment on 493c867)

File: lib/spack/spack/cmd/pkg.py
Comment (becker33) [inline lib/spack/spack/cmd/pkg.py]: Why make this restriction? Couldn't we just print each hash sequentially on its own line?

File: lib/spack/spack/test/data/unparse/amdfftw.txt
Comment (becker33) [inline lib/spack/spack/test/data/unparse/amdfftw.txt:2]: For all of these packages, can you replace the docstring description with a description of what about this package is particularly challenging for the unparser or what this package is testing about the unparser?

File: lib/spack/spack/test/util/package_hash.py
Comment (becker33) [inline lib/spack/spack/test/util/package_hash.py]: Are these two methods supposed to have the same name? If so it could use a comment explaining why.

File: lib/spack/spack/test/util/package_hash.py
Comment (becker33) [inline lib/spack/spack/test/util/package_hash.py:406]: Can you explain a bit why this is the case? I don't think I can adequately review this block without a quick chat about why it is the way it is.

File: lib/spack/external/spack_astunparse/unparser.py
Comment (cosmicexplorer) [reply lib/spack/external/spack_astunparse/unparser.py]: I can read it when looking at the PR "conversation" tab, so I believe the answer is yes!

File: lib/spack/spack/test/data/unparse/README.md
Comment (cosmicexplorer) [inline lib/spack/spack/test/data/unparse/README.md]: This seems like a good workaround. Is there any documentation about spack checkers we might want to add this as a note to?

File: lib/spack/spack/package.py
Comment (cosmicexplorer) [inline lib/spack/spack/package.py]: Why is this imported here if the rest of the changes in this file do not use that import? Is it getting magically invoked somehow?

File: lib/spack/spack/directives.py
Comment (cosmicexplorer) [inline lib/spack/spack/directives.py:740]: This is a good change!

File: lib/spack/spack/directives.py
Comment (cosmicexplorer) [inline lib/spack/spack/directives.py]:
```suggestion
#: Names of possible directives. This list is populated elsewhere in the file and then added to `__all__` at the bottom.
```

File: lib/spack/spack/test/data/unparse/README.md
Comment (cosmicexplorer) [reply lib/spack/spack/test/data/unparse/README.md]: Do I understand correctly that the content of these files can be regenerated at any time by calling `spack pkg source <package>`? Could we add a small note on that in this file so someone who wants to add an additional test package later gets how to do that?

File: lib/spack/spack/cmd/pkg.py
Comment (cosmicexplorer) [inline lib/spack/spack/cmd/pkg.py]: Recommend flipping the logical condition here since it can sometimes be confusing to read `if not x` when both sides of `if`/`else` are used. This might look like:

```python
    if args.canonical:
        content = ph.canonical_source(spec)
        sys.stdout.write(content)
        return

    # regular source dump -- just get the package and print its contents
    filename = spack.repo.patch.filename_for_package_name(spec.name)
    if sys.stdout.isatty():
        tty.msg("%s:" % filename)
    with open(filename) as f:
        sys.stdout.write(f.read())
```

However, it's also not clear to me why we would only `tty.msg()` the `filename` when we do a non-`canonical` source dump. To remove some of this logic, I might recommend the following: 

```python
    filename = spack.repo.path.filename_for_package_name(spec.name)
    if sys.stdout.isatty():
        tty.msg("%s:" % filename)

    if args.canonical:
        content = ph.canonical_source(spec)
    else:
        with open(filename) as f:
            content = f.read() # we might have an llnl.util.filesystem method for this

    sys.stdout.write(content)
```

File: lib/spack/spack/test/cmd/pkg.py
Comment (cosmicexplorer) [inline lib/spack/spack/test/cmd/pkg.py:264]: I get worried when I see `assert ... not in ...`, since `not in` could be true if it spit out the completely wrong info as well. Would it be too brittle if we took the output of the `spack pkg source --canonical multimethod@1.0` command as another `.txt` file and simply compared it directly to the output here?

File: lib/spack/spack/test/cmd/pkg.py
Comment (cosmicexplorer) [inline lib/spack/spack/test/cmd/pkg.py]: Is this just testing that the `spack pkg hash multimethod` command doesn't fail? Could we test against its actual hash now that the hash is consistent?

File: lib/spack/spack/test/package_sanity.py
Comment (cosmicexplorer) [inline lib/spack/spack/test/package_sanity.py:79]: Is this unfinished code? I would assume we'd use an `assert` here in order to make sure this test fails?

File: lib/spack/spack/util/package_hash.py
Comment (cosmicexplorer) [inline lib/spack/spack/util/package_hash.py]:
```suggestion
    This removes *all* strings that aren't on the RHS of an assignment statement from the body of functions, classes,
```

File: lib/spack/spack/util/package_hash.py
Comment (cosmicexplorer) [inline lib/spack/spack/util/package_hash.py]:
```suggestion
    possibly affect package behavior in the package hash, and we want to exclude things
```

Comment (cosmicexplorer) [reply lib/spack/spack/util/package_hash.py]: This is a fantastically helpful explanation.

File: lib/spack/spack/util/package_hash.py
Comment (cosmicexplorer) [inline lib/spack/spack/util/package_hash.py:267]: Recommend turning this into an `Exception` subclass before merging.

File: lib/spack/spack/util/package_hash.py
Comment (cosmicexplorer) [inline lib/spack/spack/util/package_hash.py]:
```suggestion
    This function is used to determine whether a spec needs a rebuild when a
```

File: lib/spack/spack/util/unparse/unparser.py
Comment (cosmicexplorer) [inline lib/spack/spack/util/unparse/unparser.py:16]: Possibly a job for #21720.

File: lib/spack/spack/util/unparse/unparser.py
Comment (cosmicexplorer) [inline lib/spack/spack/util/unparse/unparser.py:22]: Is there documentation for this behavior in the `ast` or `unparse` modules anywhere?

File: lib/spack/spack/util/unparse/unparser.py
Comment (cosmicexplorer) [inline lib/spack/spack/util/unparse/unparser.py:110]: Good docstring!

**Review by cosmicexplorer — DISMISSED** [review]
This is awesome!!!!!!! I'm marking approve because I would accept this as-is, but will pay attention to any discussion on my review comments.

File: lib/spack/spack/test/data/unparse/README.md
Comment (tgamblin) [reply lib/spack/spack/test/data/unparse/README.md]: They're just copies of packages -- they need to remain static so their hashes don't change.

File: lib/spack/spack/package.py
Comment (tgamblin) [reply lib/spack/spack/package.py]: I am very confused by this -- it seems like flake8 is overlooking it.

File: lib/spack/spack/util/unparse/unparser.py
Comment (tgamblin) [reply lib/spack/spack/util/unparse/unparser.py:22]: No, but unparser.py is basically an external -- I'm hesitant to change much more about it.

File: lib/spack/spack/util/unparse/unparser.py
Comment (tgamblin) [reply lib/spack/spack/util/unparse/unparser.py:16]: yeah we should probably resurrect that.

File: lib/spack/spack/util/package_hash.py
Comment (tgamblin) [reply lib/spack/spack/util/package_hash.py:267]: Why?  If this isn't true, it's a logic error that should come out in testing, not something that should be dealt with at runtime by any client code.

File: lib/spack/spack/package.py
Comment (tgamblin) [reply lib/spack/spack/package.py]: I just removed this.

File: lib/spack/spack/cmd/pkg.py
Comment (tgamblin) [reply lib/spack/spack/cmd/pkg.py]: ok did this.

File: lib/spack/spack/test/cmd/pkg.py
Comment (tgamblin) [reply lib/spack/spack/test/cmd/pkg.py:264]: I think that's pretty brittle, and there are a lot of `in` tests in addition to `not in` tests here.

File: lib/spack/spack/test/cmd/pkg.py
Comment (tgamblin) [reply lib/spack/spack/test/cmd/pkg.py]: The hash is only consistent if the package doesn't change.  So testing that would require another argument to grab the source from some alternative and unchanging place, but we already have other tests that do that... I didn't think it was worth the trouble.

File: lib/spack/spack/util/package_hash.py
Comment (becker33) [inline lib/spack/spack/util/package_hash.py:187]: I'm a little concerned with how this will play with multimethods on inherited packages. Can you make sure to add a test for parsing a package that inherits from a package with multimethod and overrides, and another for adding a condition to the multimethod?

File: lib/spack/spack/util/package_hash.py
Comment (becker33) [inline lib/spack/spack/util/package_hash.py]: Can we clean up the calling convention for these? It looks like `RemoveDirectives.visit` and `TagMultiMethods.visit` operate in place, but `RemoveDocstrings.visit` and `ResolveMultiMethods.visit` return a copy

**Review by becker33 — CHANGES_REQUESTED** [review]

File: lib/spack/spack/cmd/pkg.py
Comment (tgamblin) [reply lib/spack/spack/cmd/pkg.py]: Yep good point.  Fixed.

File: lib/spack/spack/test/data/unparse/amdfftw.txt
Comment (tgamblin) [reply lib/spack/spack/test/data/unparse/amdfftw.txt:2]: done

File: lib/spack/spack/test/util/package_hash.py
Comment (tgamblin) [reply lib/spack/spack/test/util/package_hash.py]: it actually doesn't matter for this test but they don't need to be named the same thing and probably shouldn't be.  I renamed them method1 and method2 and added a check to ensure that they're still in the unparsed output.

File: lib/spack/spack/test/util/package_hash.py
Comment (tgamblin) [reply lib/spack/spack/test/util/package_hash.py:406]: Is the [explanation in package_hash.py](https://github.com/spack/spack/pull/28156/files#diff-c0e9963b0d8dcee67b8fc64f623170a84ac7c13892f504c95106e8d1c4cdaf13R136-R178) sufficient?

The reason you need both is because you don't know until runtime whether all the `@when`'s on  `THREE` will end up being true, b/c you don't know whether you'll end up running the build on Darwin or not.  If you're on Darwin, THREE wins.  otherwise default does b/c TWO, FOUR, and FIVE are false, and the default is preserved in case nothing wins.

File: lib/spack/spack/util/package_hash.py
Comment (tgamblin) [reply lib/spack/spack/util/package_hash.py:284]: Yep this is a good idea.

File: lib/spack/spack/test/data/unparse/README.md
Comment (cosmicexplorer) [reply lib/spack/spack/test/data/unparse/README.md]: Ok, that sounds reasonable!

File: lib/spack/spack/test/cmd/pkg.py
Comment (cosmicexplorer) [reply lib/spack/spack/test/cmd/pkg.py]: Ok, great!

File: lib/spack/spack/test/cmd/pkg.py
Comment (cosmicexplorer) [reply lib/spack/spack/test/cmd/pkg.py:264]: Ok great!

File: lib/spack/spack/util/package_hash.py
Comment (cosmicexplorer) [reply lib/spack/spack/util/package_hash.py:267]: Ok, I misunderstood the intent of "there is a problem" in the comment! I think this is fine as-is then.

File: lib/spack/spack/util/unparse/unparser.py
Comment (cosmicexplorer) [reply lib/spack/spack/util/unparse/unparser.py:16]: you got it boss

File: lib/spack/spack/util/unparse/unparser.py
Comment (cosmicexplorer) [reply lib/spack/spack/util/unparse/unparser.py:22]: Oh, right! Thanks--sounds great to keep as-is!

File: lib/spack/spack/util/package_hash.py
Comment (tgamblin) [reply lib/spack/spack/util/package_hash.py]: The tagger is just a visitor and doesn't modify the tree, so I didn't tweak that one, but the other two should be assignments because they are `NodeTransformer` subclasses.  I added some comments to make this clearer

File: lib/spack/spack/util/package_hash.py
Comment (tgamblin) [reply lib/spack/spack/util/package_hash.py:187]: So, at the moment there are no inherited multimethods in Spack mainline (either in `build_systems` or in any package that's imported from another), and base packages aren't included in the package hash at all.

There is a question of where to draw the line for the package hash.  I think eventually we should consider making specs even finer-grained and include the spack version or commit in their hashes.  At the same time we might want to include all bases up to but not including `PackageBase`, but then that overlaps with the spack hash in terms of what's represented in a spec.  We could include anything that's *not* in Spack itself, I suppose.

The multimethod stuff is separate because whether we preserve subclass multi methods can depend on whether we're overriding a multimethod from a parent class.  I think that analysis should get factored in, but currently we don't have that information in the AST analysis -- we would need to look up the parent class and parse its AST, as well.

So ultimately I think it's important to handle all this, but I think we should do it in a separate PR.

**Comment by tgamblin** [comment]
@becker33 this is ready for another look, and I have some questions.

**Comment by tgamblin** [comment]
I guess the last question is what to do with `unparser.py`.  Currently its coverage is low because we're not running the full unparse test suite on it.  But, it's heavily modified and has code from `astunparse` and a bunch of backports from upstream Python.

I think maybe it should be its own repo in the spack org, with its own tests to ensure that it is consistent with upstream `ast.unparse()`.  That would probably let us test it more easily, and it would let us monitor its coverage more easily.  I think other projects might benefit from a canonical unparser, too.  Any thoughts on this?

**Comment by tgamblin** [comment]
For now I just added unit tests to ensure that we're covering all of the unparser logic.  I think we can factor this out later.

I looked at packages and discovered that we weren't handling `with when()`, `for`, `if`, etc. or some attributes like `maintainers` properly.  I added `maintainers` and `tags` to ignored attributes, and the unparser now also removes nested logic if it contains only directives.  I think this is a much better package hash now.

@becker33 can you take one more look?

File: lib/spack/spack/util/package_hash.py
Comment (scottwittenburg) [reply lib/spack/spack/util/package_hash.py]: I agree, this docstring really helped me understand the issue with multimethods.

**Review by scottwittenburg — COMMENTED** [review]
I won't up or down this, but I did look through it deeply enough to gain a better understanding of how the package was computed before, and why this is an improvement.  Thanks for the detailed PR description and some great docstrings discussing the finer points.

As you pointed out somewhere, each gitlab CI pipeline stack is only run with a single version of python, which is baked into the container used to concretize the stack, so maybe this won't reduce the number of rebuilds for that reason.  But more consistent package hashing seems like an important prerequisite to using the full hash everywhere.

File: lib/spack/spack/test/cmd/pkg.py
Comment (becker33) [inline lib/spack/spack/test/cmd/pkg.py]:
```suggestion
    assert len(output) == 2 and all(len(elt) == 32 for elt in output)
```
Otherwise python will recognize that the generator is of non-zero length, and therefore treat it as true-ish even if all elements are `False`.

File: lib/spack/spack/test/cmd/pkg.py
Comment (becker33) [inline lib/spack/spack/test/cmd/pkg.py]:
```suggestion
    assert len(output) == 1 and all(len(elt) == 32 for elt in output)
```

File: lib/spack/spack/util/package_hash.py
Comment (becker33) [reply lib/spack/spack/util/package_hash.py:187]: I think it's okay to push this to another PR, but that does mean we will need to have another hash-breaking PR in the future.

**Review by becker33 — DISMISSED** [review]
I think this has addressed all my review comments, excepting the issue we pushed to a future PR regarding multimethods inherited from another package in the MRO.

Two very minor issues with the tests, should be able to just commit the suggestions below.

File: lib/spack/spack/test/util/package_hash.py
Comment (becker33) [reply lib/spack/spack/test/util/package_hash.py:406]: Yes, I think that's sufficient. It's frustrating that we can't fully canonicalize -- I suspect that most of these packages could have `'platform=darwin'` instead of `sys.platform == 'darwin'`, which would let us trim a little more.

Comment (tgamblin) [reply lib/spack/spack/test/util/package_hash.py:406]: While I was thinking about this, I realized that we should consider disallowing dynamic conditions like this in the future for another reason: it interferes with indexing the package metadata. We can deal with that later, too.

File: lib/spack/spack/util/package_hash.py
Comment (tgamblin) [reply lib/spack/spack/util/package_hash.py:187]: I think this becomes increasingly less impactful the sooner we make reuse the default.

**Comment by tgamblin** [comment]
@becker33 all changes are committed and squashed into prior commits.  Thanks!

File: lib/spack/spack/util/package_hash.py
Comment (becker33) [reply lib/spack/spack/util/package_hash.py:187]: It will still cause a CI backlog at least, even if users are insulated from it

**Review by becker33 — DISMISSED** [review]

**Comment by tgamblin** [comment]
@spackbot run pipeline

**Comment by tgamblin** [comment]
@spackbot run pipeline

**Review by alalazo — APPROVED** [review]

**Pull request merged**
